File: Ejercicios/Ejercicio3.py
# 3. Create an university enrollment system with the following characteristics:
# * 		(listo)The system has a login with a username and password.
# * 		(listo)Upon logging in, a menu displays the available programs: Computer Science, Medicine, Marketing, and Arts.
# * 		The user must input their first name, last name, and chosen program.
# * 		Each program has only 5 available slots. The system will store the data of each registered user, and if it exceeds the limit, it should display a message indicating the program is unavailable.
# * 		(listo)If login information is incorrect three times, the system should be locked.
# * 		The user must choose a campus from three cities: London, Manchester, Liverpool.
# * 		In London, there is 1 slot per program; in Manchester, there are 3 slots per program, and in Liverpool, there is 1 slot per program.
# * 		If the user selects a program at a campus that has no available slots, the system should display the option to enroll in the program at another campus.
import sys

class account:    
    def __init__(self):
        self.user = "a"
        self.password = "1"
        self.students ={}
        self.programs = {
            "Computer Science": {"London": 1, "Manchester": 3, "Liverpool": 1},
            "Medicine": {"London": 1, "Manchester": 3, "Liverpool": 1},
            "Marketing": {"London": 1, "Manchester": 3, "Liverpool": 1},
            "Arts": {"London": 1, "Manchester": 3, "Liverpool": 1}
        }
    


class login(account):
    def __init__(self):
        super().__init__()
        self.failed_attempts = 0
    
    def verify_user(self, input_user, input_password):
        if input_user == self.user and input_password == self.password:
            print("Successful Login!!")

            return True
            
        else:
            print("Incorrect Password!")
            self.failed_attempts += 1
            if self.failed_attempts >= 3:
                print("error number 3. The program will close")
                sys.exit()
            return False
        

class menu(account):
    def select(self):
        

        
        print("possible available programs!!!")
        print("-------------------------------")
        print("1. Computer science")
        print("2. Medicine")
        print("3. Marketing")
        print("4. Arts")
        print("5. Exit")
        print("--------------------------------")
    
    
        self.opcion = int(input(f"Enter option: "))
        return self.opcion



class actions(menu):
    
    def register(self, program):
        first_name = input("Enter your first name: ")
        last_name = input("Enter your last name: ")
        campus = input ("Choose a campus (London, Manchester, Liverpool): ")
        if self.programs[program][campus]>0:
            self.programs[program][campus] -= 1
            self.students[first_name + " " + last_name] = {"program": program, "campus": campus}
            print("Registration successful!")
        else:
            print("No available slots in this program at the chosen campus. Please choose another campus or program.")
    
    def computer_science(self):
        self.register("Computer Science")        
        print("Computer Science")
        
    def medicine(self):
        self.register("Medicine")    
        print("Medicine")
   
    def marketing(self):
        self.register("Marketing")    
        print("Marketing ")
        
   
    def arts(self):
        self.register("Arts")    
        print("Arts")
   
    def exit(self):
        print("exit")
        sys.exit()

    def numbers(self, opcion):
        if opcion == 1:
            self.computer_science()
        elif opcion == 2:
            self.medicine()
        elif opcion == 3:
            self.marketing()
        elif opcion == 4:
            self.arts()
        elif opcion == 5:
            self.exit()
        else:
           print("Invalid option")






if __name__ == "__main__":
    login_instance = login()

    logging = False
    while not logging:
        input_user = input("Enter your username: ")
        input_password = input("Enter your password: ")
        logging = login_instance.verify_user(input_user, input_password)
    menu_instance = menu()
    actions_instance = actions()
    while True:
        opcion = menu_instance.select()
        actions_instance.numbers(opcion)

File: Ejercicio3/ejercicio3.js
// 3. Create an university enrollment system with the following characteristics:
// * 	The system has a login with a username and password.
// * 	Upon logging in, a menu displays the available programs: Computer Science, Medicine, Marketing, and Arts.
// * 	The user must input their first name, last name, and chosen program.
// * 	Each program has only 5 available slots. The system will store the data of each registered user, and if it exceeds the limit, it should display a message indicating the program is unavailable.
// * 	If login information is incorrect three times, the system should be locked.
// * 	The user must choose a campus from three cities: London, Manchester, Liverpool.
// * 	In London, there is 1 slot per program; in Manchester, there are 3 slots per program, and in Liverpool, there is 1 slot per program.
// * 	If the user selects a program at a campus that has no available slots, the system should display the option to enroll in the program at another campus.
const readline = require("readline");
let name;
let lastname;
let Program;
const users = [
  { username: "user1", password: "pass123" },
  { username: "user2", password: "pass456" },
  { username: "user3", password: "pass789" },
];
const programs = {
    ComputerScience: {
      totalStudents: 0,
      campuses: {
        London: [],
        Manchester: [],
        Liverpool: [],
      },
    },
    Medicine: {
      totalStudents: 0,
      campuses: {
        London: [],
        Manchester: [],
        Liverpool: [],
      },
    },
    Marketing: {
      totalStudents: 0,
      campuses: {
        London: [],
        Manchester: [],
        Liverpool: [],
      },
    },
    Arts: {
      totalStudents: 0,
      campuses: {
        London: [],
        Manchester: [],
        Liverpool: [],
      },
    },
  };
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
function askQuestion(question) {
  return new Promise((resolve) => {
    rl.question(question, resolve);
  });
}
async function login() {
  let username;
  let password;
  let tries = 1;

  while (tries <= 3) {
    username = await askQuestion("Enter your username: ");
    password = await askQuestion("Enter your password: ");

    const user = users.find(
      (student) =>
        student.username === username && student.password === password
    );

    if (user) {
      console.log("Login successful!");
      return true;
    } else {
      console.log("Incorrect credentials. Please try again.");
      tries++;
    }
  }

  console.log("Maximum attempts exceeded. Exiting...");
  return false;
}
async function enrollment() {
  console.log("Choose the program to enroll in");
  console.log("1. Computer Science");
  console.log("2. Medicine");
  console.log("3. Marketing");
  console.log("4. Arts");
  let NumProgram;
  try {
    NumProgram = await askQuestion("Please enter the number associated with the program: ");
    if (isNaN(Number(NumProgram)) || NumProgram < 1 || NumProgram > 4) {
      throw new Error("Invalid program number. Please enter a valid number.");
    }
  } catch (error) {
    console.log(error.message);
    await enrollment(); // Reinicia el proceso de inscripción si hay un error
    return;
  }
   name = await askQuestion("Please enter your name: ");
   lastname = await askQuestion("Please enter your lastname: ");

  switch (Number(NumProgram)) {
    case 1:
      Program = "ComputerScience";
      break;
    case 2:
      Program = "Medicine";
      break;
    case 3:
      Program = "Marketing";
      break;
    case 4:
      Program = "Arts";
      break;

    default:
      break;
  }

  if (programs[Program].totalStudents < 5) {
    console.log(Program);
    await chooseCampus();
  }else{
    console.log(`Sorry, there is no available slot for ${Program}.`);

  }
}
async function chooseCampus() {
    let campus;
    let campusFull = true;
  
    while (campusFull) {
      console.log("Choose a campus:");
      console.log("1. London");
      console.log("2. Manchester");
      console.log("3. Liverpool");
  
      let numCampus;
      try {
        numCampus = await askQuestion("Enter the number of your chosen campus: ");
        if (isNaN(Number(numCampus)) || numCampus < 1 || numCampus > 3) {
          throw new Error("Invalid campus number. Please enter a valid number.");
        }
      } catch (error) {
        console.log(error.message);
        continue;
      }
      
      switch (Number(numCampus)) {
        case 1:
          campus = "London";
          break;
        case 2:
          campus = "Manchester";
          break;
        case 3:
          campus = "Liverpool";
          break;
        default:
          break;
      }
  
      if (campus === "London" || campus === "Liverpool") {
        if (programs[Program].campuses[campus].length < 1) {
          campusFull = false;
        } else {
          console.log("The selected campus is full. Please choose another campus.");
        }
      } else if (campus === "Manchester") {
        if (programs[Program].campuses[campus].length < 3) {
            campusFull = false;
          } else {
            console.log("The selected campus is full. Please choose another campus.");
          }
        
      }
    }
  
    console.log(`${Program} ${campus}`);
    await registerStudent(Program, campus);
  }
async function registerStudent(program, campus) {
    const studentName = `${name} ${lastname}`;
    programs[program].totalStudents++;
    programs[program].campuses[campus].push(studentName)
    console.log(`
    Student Information:
    Name: ${name}
    Lastname: ${lastname}
    Program: ${program}
    Campus: ${campus}
    Enrollment successful!
  `);
  const continueEnrollment = await askQuestion("Do you want to continue enrollment? (yes/no): ");
  if (continueEnrollment.toLowerCase() === 'yes') {
    await enrollment();
  } else {
    rl.close();
  }
    
    
}
async function main() {
  const loggedIn = await login();
  if (loggedIn) {
    await enrollment();
  }
  rl.close();
}
main();
